rust/src/parallel_map.rs

```diff
@@ -0,0 +1,206 @@
+where
+    I: Iterator,
+    I::Item: Send,
+    T: Send,
+{
+    fn drop(&mut self) {
+        // Send end of communication to all threads.
+        for communication in &self.communication {
+            let _ = communication.send.send(None);
+        }
+        self.communication.clear();
+
+        // Join all threads.
+        while let Some(handle) = self.handles.pop() {
+            let _ = handle.join();
+        }
+    }
+}
+
+/// Map `fun` on `iter` using `threads` threads. This can be used with any Iterator and returns an
+/// Iterator.
+///
+/// Example
+/// ```rust
+/// let s: i64 = parallel_map(|x| 2 * x, (1 .. 6).into_iter(), 15).sum();
+/// assert_eq!(s, 30);
+/// ```
+pub fn parallel_map<I, T>(fun: fn(I::Item) -> T, mut iter: I, threads: usize) -> ParallelMap<I, T>
+where
+    I: Iterator,
+    I::Item: Clone + Send + 'static,
+    T: Send + 'static,
+{
+    // At least some workers.
+    assert!(threads >= 1);
+
+    // Start the threads and send an item to each of them.
+    let mut communication = Vec::new();
+    let mut handles = Vec::new();
+    for t in 0 .. threads {
+        // Next task for the thread.
+        let next_task = match iter.next() {
+            None => break, // Not creating a thread for nothing.
+            Some(next_task) => Some(next_task),
+        };
+
+        // Create the thread.
+        let (par_map, thread) = ThreadCommunication::<I::Item, T>::new_pair();
+        communication.push(par_map);
+        let handle = std::thread::spawn(move || loop {
+            match thread.receive.recv() {
+                Ok(optional_task) => match optional_task {
+                    None => return,
+                    Some(task) => match thread.send.send(Some((fun)(task))) {
+                        Ok(()) => (),
+                        Err(_) => return,
+                    },
+                },
+                Err(_) => return,
+            };
+        });
+        handles.push(handle);
+
+        // Send the task.
+        let _ = communication[t].send.send(next_task);
+    }
+
+    ParallelMap { now: 0, iter, communication, handles }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    fn add_one(x: i32) -> i32 {
+        x + 1
+    }
+
+    #[test]
+    fn plus_one() {
+        for iterator_length in 0 .. 50 {
+            for threads in 1 .. 20 {
+                let mut iterations = 0;
+                for (i, res) in parallel_map(add_one, 0 .. iterator_length, threads).enumerate() {
+                    // The value is correct and deterministic.
+                    assert_eq!((i + 1) as i32, res);
+                    iterations += 1;
+                }
+                // The resulting iterator is neither shorter or longer.
+                assert_eq!(iterations, iterator_length);
+            }
+        }
+    }
+
+    fn sleepy_add_one(x: i32) -> i32 {
+        std::thread::sleep(std::time::Duration::from_millis(10));
+        x + 1
+    }
+
+    #[test]
+    fn test_doc() {
```

Comment:
You don't need to do that. It's automatic. Running `cargo test` will compile and execute the doc examples.

Reply:
It should. Bu I get this:
```bash
cargo test --doc
warning: doc tests are not supported for crate type(s) `cdylib` in package `sedpack_rs`
error: no library targets found in package `sedpack_rs`
```

Reply:
Fantastic, I was missing the `rlib`. And `pub mod` it is *shrug*.